relational_operators/BuildHashOperator.cpp

```diff
@@ -56,68 +54,174 @@ class TupleReferenceGenerator {
+                                                           input_relation_block_ids_[num_workorders_generated_],
+                                                           hash_table,
+                                                           storage_manager),
+                                    op_index_);
+      ++num_workorders_generated_;
+    }
+  }
+}
 
-  JoinHashTable *hash_table = query_context->getJoinHashTable(hash_table_index_);
+void BuildHashOperator::addWorkOrdersUsingPartitionedInput(WorkOrdersContainer *container,
+                                                           QueryContext *query_context,
+                                                           StorageManager *storage_manager) {
+  const std::size_t num_partitions =
+      input_relation_.getPartitionScheme().getPartitionSchemeHeader().getNumPartitions();
+  JoinHashTable *hash_table = query_context->getJoinHashTable(hash_table_group_index_);
   if (input_relation_is_stored_) {
-    if (!started_) {
-      for (const block_id input_block_id : input_relation_block_ids_) {
+    for (std::size_t part_id = 0; part_id < num_partitions; ++part_id) {
+      for (const block_id input_block_id : input_relation_block_ids_in_partition_[part_id]) {
+        container->addNormalWorkOrder(new BuildHashWorkOrder(input_relation_,
+                                                             join_key_attributes_,
+                                                             any_join_key_attributes_nullable_,
+                                                             input_block_id,
+                                                             hash_table,
+                                                             storage_manager),
+                                      op_index_);
+      }
+    }
+  } else {
+    for (std::size_t part_id = 0; part_id < num_partitions; ++part_id) {
+      while (num_workorders_generated_in_partition_[part_id] <
+             input_relation_block_ids_in_partition_[part_id].size()) {
+        block_id block_in_partition =
+            input_relation_block_ids_in_partition_[part_id][num_workorders_generated_in_partition_[part_id]];
+        container->addNormalWorkOrder(new BuildHashWorkOrder(input_relation_,
+                                                             join_key_attributes_,
+                                                             any_join_key_attributes_nullable_,
+                                                             block_in_partition,
+                                                             hash_table,
+                                                             storage_manager),
+                                      op_index_);
+        ++num_workorders_generated_in_partition_[part_id];
+      }
+    }
+  }
+}
+
+#ifdef QUICKSTEP_HAVE_LIBNUMA
+void BuildHashOperator::addPartitionAwareWorkOrders(WorkOrdersContainer *container,
```

Comment:
Could we rename to `addNumaAwareWorkOrders`, as in the distributed case we would be partition aware, but not `NUMA` aware.